components/raftstore/src/store/peer.rs

```diff
@@ -477,17 +479,167 @@ pub struct ReadyResult {
 /// 6. After the plan steps are all applied, exit force leader state
 pub enum ForceLeaderState {
     WaitTicks {
+        shared_state: UnsafeRecoveryForceLeaderSharedState,
         failed_stores: HashSet<u64>,
         ticks: usize,
     },
     PreForceLeader {
+        shared_state: UnsafeRecoveryForceLeaderSharedState,
         failed_stores: HashSet<u64>,
     },
     ForceLeader {
         failed_stores: HashSet<u64>,
     },
 }
 
+// Following shared states are used while reporting to PD for unsafe recovery and shared among
+// all the regions per their life cycle.
+// The work flow is like:
+//     1. report phase
+//            start_unsafe_recovery_report
+//            -> broadcast wait-apply commands
+//            -> wait for all the peers' apply indices meet their targets
+//            -> broadcast fill out report commands
+//            -> wait for all the peers fill out the reports for themsleves
+//            -> send a store report (through store heartbeat)
+//     2. force leader phase
+//            dispatch force leader commands
+//            -> wait for all the peers that recived the command become force leader
+//            -> start_unsafe_recovry_report
+//     3. plan execution phase
+//            dispatch recovery plans
+//            -> wait for all the creates, deletes and demotes to finish, for the demotes,
+//               procedures are:
+//                   -> exit residual joint state
+//                   -> demote failed voters, and promote self to be a voter if it is a learner
+//                   -> exit joint state
+//            -> start_unsafe_recovery_report
+#[derive(Copy, Clone, Debug)]
+pub struct UnsafeRecoveryReportContext {
+    pub id: u64,
+    pub exit_force_leader_after_reporting: bool,
+}
+// Intends to use RAII to sync unsafe recovery procedures between peers, in addition to that,
+// it uses a closure to avoid having a raft router as a member variable, which is statically
+// dispatched, thus needs to propage the generics everywhere.
+pub struct InvokeClosureOnDrop(Box<dyn Fn() + Send + Sync>);
+impl fmt::Debug for InvokeClosureOnDrop {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "InvokeClosureOnDrop")
+    }
+}
+impl Drop for InvokeClosureOnDrop {
+    fn drop(&mut self) {
+        self.0();
+    }
+}
+pub fn start_unsafe_recovery_report<EK: KvEngine, ER: RaftEngine>(
+    router: &RaftRouter<EK, ER>,
+    report_context: UnsafeRecoveryReportContext,
+) {
+    info!("Unsafe recovery, preparing report");
+    let wait_apply = UnsafeRecoveryWaitApplySharedState::new(report_context, router.clone());
+    router.broadcast_normal(|| {
```

Comment:
Do we need to broadcast this message to all peers or just to specified peers?

Reply:
This is due to the reason that commit indices and apply indices are hard to get in `work/pd.rs`. Besides, unified behavior is more readable considering this is not on the performance critical path.

Reply:
It's necessary. Split and merge are paused for the whole cluster, we need to make sure all the previous split and merge are finished.